### PR TITLE
Better Support for Guard Clauses in ThrowsCount Rule

### PR DESCRIPTION
**Pull request by amitdash291**
This PR achieves the following:
 - Make `GuardClauses` extension functions generic to support both `return` and `throw` expressions
 - Fix `ThrowsCount` scenarios related to guard clauses and add tests for the same
 - Reuse `GuardClauses` code for `ThrowsCount` rule

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ThrowsCountSpec.kt [hunk 1]
```diff
@@ -73,8 +71,132 @@ class ThrowsCountSpec : Spek({
...
+                    }
+                    throw Exception()
+                }
+                when (x) {
+                    5 -> println("x=5")
+                    4 -> throw Exception()
+                }
+                throw Exception()
+            }
+        """
```
Comment (arturbosch) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ThrowsCountSpec.kt:129]: It looks like all test code templates need 4 extra spaces

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ThrowsCountSpec.kt [hunk 2]
```diff
@@ -73,8 +71,132 @@ class ThrowsCountSpec : Spek({
...
+                    val models = b as? Int ?: throw Exception()
+                    val position = c?.takeIf { it != -1 } ?: throw Exception()
+                    if(b !is String) {
+                        println("b is not a String")
+                        throw Exception()
+                    }
+
+                    throw Exception()
+                }
+            """
```
Comment (arturbosch) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ThrowsCountSpec.kt:188]: Here it is right

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt [hunk 3]
```diff
@@ -5,39 +5,36 @@ import org.jetbrains.kotlin.psi.KtBinaryExpression
...
 
-fun KtNamedFunction.yieldStatementsSkippingGuardClauses(): Sequence<KtExpression> = sequence {
-    var firstNonGuardFound = false
-    this@yieldStatementsSkippingGuardClauses.bodyBlockExpression?.statements?.forEach {
-        if (firstNonGuardFound) {
-            yield(it)
-        } else {
-            if (!it.isGuardClause()) {
-                firstNonGuardFound = true
+inline fun <reified T : KtExpression> KtNamedFunction.yieldStatementsSkippingGuardClauses(): Sequence<KtExpression> =
```
Comment (arturbosch) [inline detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt:11]: Should this return T instead?

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt [hunk 4]
```diff
@@ -5,39 +5,36 @@ import org.jetbrains.kotlin.psi.KtBinaryExpression
...
+fun <T : KtExpression> KtExpression.isIfConditionGuardClause(descendantExpr: T): Boolean {
+    val ifExpr = this as? KtIfExpression ?: return false
+    return ifExpr.`else` == null &&
+            descendantExpr === ifExpr.then?.lastBlockStatementOrThis()
 }
 
 fun KtExpression.isElvisOperatorGuardClause(): Boolean {
-    val elvisExpr = this.parent as? KtBinaryExpression
-    return elvisExpr?.operationToken == KtTokens.ELVIS
+    val elvisExpr = this.findDescendantOfType<KtBinaryExpression>() ?: return false
```
[the remark this answers is not in the document]
Comment (amitdash291) [reply detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt:36]: Earlier `isElvisOperatorGuardClause ` was being called as an extension for the child `return` expression, and now I have changed it to be called by the parent statement containing the `return` expression.  
Firstly it makes it generic for both `return` and `throw` expressions, as it is called on the parent. And secondly it sounds more meaningful, as the the parent statement is a guard clause (and not the child).

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt [hunk 5]
```diff
@@ -5,39 +5,36 @@ import org.jetbrains.kotlin.psi.KtBinaryExpression
...
+        var firstNonGuardFound = false
+        this@yieldStatementsSkippingGuardClauses.bodyBlockExpression?.statements?.forEach {
+            if (firstNonGuardFound) {
                 yield(it)
+            } else {
+                if (!it.isGuardClause<T>()) {
+                    firstNonGuardFound = true
+                    yield(it)
+                }
             }
```
[the remark this answers is not in the document]
Comment (amitdash291) [reply detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt]: Thanks for the suggestion, it is applied.

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt (hunk 3 above)
Comment (amitdash291) [reply detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt:11]: @arturbosch sorry if it was not clear, but the generic `T` is for the child expression which can be either `return` or `throw`. These expressions can be of the sub-types `KtReturnExpression` or `KtThrowExpression`. 
Whereas the return type of `yieldStatementsSkippingGuardClauses` method is the parent statement, which is always of the type `KtExpression`.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ThrowsCountSpec.kt (hunk 1 above)
Comment (amitdash291) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ThrowsCountSpec.kt]: @arturbosch good catch, I have fixed it.

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (amitdash291) [reply detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt]: @schalkms thanks for the suggestion, it has been applied.

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (amitdash291) [reply detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt:36]: @schalkms 
Earlier isElvisOperatorGuardClause was being called as an extension for the child return expression, and now I have modified it to be called as an extension on the parent statement (which contains the return / throw child expression)
Firstly it makes it generic for both return and throw expressions, as it is called on the parent. And secondly it sounds more meaningful, as the the parent statement is a guard clause (and not the child).

**Comment by amitdash291** [comment]
@arturbosch @schalkms 
I have a question on the code coverage check. No matter how many spek tests cases I try to add (exhaustively covering most of the code branches), it always says coverage has degraded. Any hints on how to fix this?

File: detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt (hunk 3 above)
Comment (arturbosch) [reply detekt-psi-utils/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/GuardClauses.kt:11]: Thanks for the explanation.

**Comment by arturbosch** [comment]
@amitdash291 for `GuardClauses` it seems that jacoco does not probably instrument the bytecode. Maybe some edge case for Kotlin sequence generators. Their Kotlin support is still fresh.
Sometimes we also have some problems with CI caching ^^.
PR looks fine, I'll merge it ^^

**Comment by amitdash291** [comment]
> @amitdash291 for `GuardClauses` it seems that jacoco does not probably instrument the bytecode. Maybe some edge case for Kotlin sequence generators. Their Kotlin support is still fresh.
> Sometimes we also have some problems with CI caching ^^.
> PR looks fine, I'll merge it, thanks!

@arturbosch got it, the clarification about coverage issue.
Thanks for merging this PR 👍

**Pull request merged**